lsp/gitlab_duo.lua

```diff
@@ -0,0 +1,473 @@
+end
+
+-- OAuth Device Flow
+local function device_authorization()
+  local response = curl_post(
+    config.gitlab_url .. '/oauth/authorize_device',
+    string.format('client_id=%s&scope=%s', config.client_id, config.scopes),
+    { ['Content-Type'] = 'application/x-www-form-urlencoded' }
+  )
+
+  if response.status ~= 200 then
+    vim.notify('Device authorization failed: ' .. response.status, vim.log.levels.ERROR)
+    return nil
+  end
+
+  local ok, data = pcall(vim.json.decode, response.body)
+  if not ok then
+    vim.notify('Failed to parse authorization response', vim.log.levels.ERROR)
+    return nil
+  end
+
+  return data
+end
+
+local function poll_for_token(device_code, interval, client)
+  local max_attempts = 60
+  local attempts = 0
+
+  local function poll()
+    attempts = attempts + 1
+
+    local response = curl_post(
+      config.gitlab_url .. '/oauth/token',
+      string.format(
+        'client_id=%s&device_code=%s&grant_type=urn:ietf:params:oauth:grant-type:device_code',
+        config.client_id,
+        device_code
+      ),
+      { ['Content-Type'] = 'application/x-www-form-urlencoded' }
+    )
+
+    local ok, body = pcall(vim.json.decode, response.body)
+    if not ok then
+      vim.notify('Failed to parse token response', vim.log.levels.ERROR)
+      return
+    end
+
+    if response.status == 200 and body.access_token then
+      save_token(body)
+      vim.notify('GitLab Duo authentication successful!', vim.log.levels.INFO)
+
+      -- Update LSP with new token
+      vim.schedule(function()
+        client.notify('workspace/didChangeConfiguration', {
+          settings = {
+            token = body.access_token,
+            baseUrl = config.gitlab_url,
+          },
+        })
+      end)
+      return
+    end
+
+    if body.error == 'authorization_pending' then
+      if attempts < max_attempts then
+        vim.defer_fn(poll, interval * 1000)
+      else
+        vim.notify('Authorization timed out', vim.log.levels.ERROR)
+      end
+    elseif body.error == 'slow_down' then
+      vim.defer_fn(poll, (interval + 5) * 1000)
+    elseif body.error == 'access_denied' then
+      vim.notify('Authorization denied', vim.log.levels.ERROR)
+    elseif body.error == 'expired_token' then
+      vim.notify('Device code expired. Please run :LspGitLabDuoSignIn again', vim.log.levels.ERROR)
+    else
+      vim.notify('OAuth error: ' .. (body.error or 'unknown'), vim.log.levels.ERROR)
+    end
+  end
+
+  poll()
+end
+
+---@param bufnr integer
+---@param client vim.lsp.Client
+local function sign_in(_bufnr, client)
+  vim.notify('Starting GitLab device authorization...', vim.log.levels.INFO)
+
+  local auth_data = device_authorization()
+  if not auth_data then
+    return
+  end
+
+  local message = string.format(
+    'GitLab Duo Authorization\n\n'
+      .. '1. Open: %s\n'
+      .. '2. Enter code: %s\n\n'
+      .. 'Code copied to clipboard. Waiting for authorization...',
```

Comment:
nit: can just a single `[[...]]` string with `vim.text.indent(0,...)` to dedent the text.